Implemented some missing operation for the NAG Fortran Compiler.

matcross opened the pull request:
Not sure whether you need a specific test case with this (since it's a proprietary compiler); here's a shell script I used for testing:

```sh
#!/bin/sh
set -eux

pdir=nagfor_proj
bdir=build

export FC=nagfor

mkdir -p "${pdir}"
cd "${pdir}"

main_file="prog.f90"
cat <<EOF > "${main_file}"
Program p
  Interface
    Function f(i)
      Integer, Intent (In) :: i
      Integer :: f
    End Function
  End Interface
  If (f(11)/=53) Stop 'FAIL'
  Print *, 'ok'
End Program
EOF

lib_file="f.f90"
cat <<EOF > "${lib_file}"
Function f(i)
  Integer, Intent (In) :: i
  Integer :: f
  f = i + 42
End Function
EOF

cat <<EOF > meson.build
project('nagfor test', 'fortran', meson_version: '>=0.59.99')
both_libs = both_libraries('mylib', '${lib_file}')
ste = executable('stprog', '${main_file}', link_with: both_libs.get_static_lib())
she = executable('shprog', '${main_file}', link_with: both_libs.get_shared_lib())
test('static_exe', ste)
test('shared_exe', she)
EOF

rm -rf "${bdir}"

meson setup "${bdir}"
meson compile -C "${bdir}"
meson test -C "${bdir}"
```

eli-schwartz commented on mesonbuild/compilers/fortran.py:
```diff
@@ -494,12 +494,24 @@ def __init__(self, exelist: T.List[str], version: str, for_machine: MachineChoic
         self.id = 'nagfor'
+        self.warn_args = {
+            '0': ['-w=all'],
+            '1': [],
```
So I take it that it reports all warnings by default?

matcross replied:
Yes that's correct.
(It also always reports a banner, which is not so great in these kinds of build systems. The forthcoming 7.1 release will add a -quiet option, hence the get_always_args change.)

dcbaker replied:
0 means no warnings, and 3 means "even useless warnings", i think

matcross replied:
Part of the nagfor design is as a linter in a way; (most) warnings are on by default. To switch off you add "-w=all: suppresses all warning messages". There could be (later) scope for adding at level 3 "-info: Request output of information messages. The default is to suppress these messages" if desired.

dcbaker replied:
Ah, okay. This is the opposite of GCC/clang where -Wall turns on a lot of warnings. Could you add a comment to the code so someone will meaning doesn't break this down the road

eli-schwartz replied:
In gcc the lowercase -w is for disabling warnings. Syntax is a bit different but the capitalization is a point of commonality.

eli-schwartz reviewed (commented):
A test would be good, it just needs to skip if nagfor is not available.

Actually... can you update run_project_tests.py -> skip_fortran to detect this compiler and run the fortran unittests in the first place?

We may not be able to test it in CI, but presumably people who have a copy can use that locally to do sanity checks on meson's support.

dcbaker commented:
You can also do `FC=nagfor ./run_project_tests.py --only fortran`, which will run just the fortran project tests. Done of them may not with with nagfor and you can add skip conditions if that makes sense. The mixed language tests may not work, for example

dcbaker commented on mesonbuild/linkers/linkers.py:
```diff
@@ -1045,6 +1045,36 @@ class QualcommLLVMDynamicLinker(LLVMDynamicLinker):
+        if build_rpath != '':
+            all_paths.add(build_rpath)
+        for rp in all_paths:
+            args.extend(self._apply_prefix('-Wl,-Wl,,-rpath,,' + rp))
```
This really needs -Wl,-Wl?

Also, isn't apply_prefix supposed to add the -Wl, for you?

Sorry I'm on my phone so reading other video is kinda hard

matcross replied:
Yes the nagfor compiler driver assembles its correct list of runtime objects and calls gcc as 'linker', therefore to pass options again to ld you need double indirection

"
-Wl,option
    Pass option directly to the host C compiler when linking (producing the executable). Multiple options may be specified in a single -Wl, option by separating them with commas. A comma may be included in an option by repeating it, e.g. -Wl,-filelist=file1,,file2,,file3 becomes the linker option -filelist=file1,file2,file3.
"

The compiler class doesn't have a prefix defined at the moment if I understand correctly (because of this extra indirection needed?), but if you advise that it should then I'll set it up that way.

dcbaker replied:
Ah, no that's fine. It's just an unusual setup. Rustc also uses GCC/clang as a linker, but it handles the correct prefixes itself. This probably deserves a comment as well

matcross commented:
> You can also do `FC=nagfor ./run_project_tests.py --only fortran`, which will run just the fortran project tests. Done of them may not with with nagfor and you can add skip conditions if that makes sense. The mixed language tests may not work, for example

Thanks for this pointer. (I believe I've now addressed the points raised above.)

I get three failures, output is pasted below.

'fortran: 9 cpp' should be expected to work, but linking needs to be done using the nagfor driver; i.e. instead of

```
cc  -o cfort cfort.p/main.c.o cfort.p/fortran.f.o -Wl,--as-needed -Wl,--no-undefined
```

use

```
nagfor -o cfort cfort.p/main.c.o cfort.p/fortran.f.o
```

but I'm not sure how to get this to occur.

Then for failures 'fortran: 15 include' and 'fortran: 21 install static (default_library=static)', I wonder if these are because of https://gitlab.kitware.com/cmake/cmake/-/issues/21398 ?

```
Meson build system 0.59.99 Project Tests
Using python 3.9.6 (default, Jul 16 2021, 00:00:00) 

host machine compilers

c      : [gcc]      cc (gcc 11.1.1 "cc (GCC) 11.1.1 20210531 (Red Hat 11.1.1-3)")
cpp    : [gcc]      c++ (gcc 11.1.1 "c++ (GCC) 11.1.1 20210531 (Red Hat 11.1.1-3)")
cs     : [not found]
cuda   : [not found]
cython : [not found]
d      : [not found]
fortran: [nagfor]   nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
java   : [not found]
objc   : [gcc]      cc (gcc 11.1.1)
objcpp : [gcc]      c++ (gcc 11.1.1)
rust   : [not found]
swift  : [not found]
vala   : [not found]

tools

ninja  : /usr/bin/ninja (1.10.2)
cmake  : /usr/bin/cmake (3.20.5)
hotdoc : not found

Checking that configuring works...
Checking that introspect works...
Checking that building works...
Checking that testing works...
Checking that installing works...

Running tests with 32 workers

Running fortran tests.

 [SUCCESS]   fortran: 1 basic
 [SUCCESS]   fortran: 2 modules
 [SUCCESS]   fortran: 3 module procedure
 [SUCCESS]   fortran: 4 self dependency
 [SUCCESS]   fortran: 5 static
 [SUCCESS]   fortran: 6 dynamic
 [SUCCESS]   fortran: 7 generated
 [SUCCESS]   fortran: 8 module names
  [ERROR]    fortran: 9 cpp
During: build
Reason: Compiling source code failed.
 [SUCCESS]   fortran: 10 find library
 [SUCCESS]   fortran: 11 compiles links runs
 [SUCCESS]   fortran: 12 submodule
 [SUCCESS]   fortran: 13 coarray
 [SUCCESS]   fortran: 14 fortran links c
  [ERROR]    fortran: 15 include
During: configure
Reason: Generating the build system failed.
 [SUCCESS]   fortran: 16 openmp
 [SUCCESS]   fortran: 17 add_languages
 [SUCCESS]   fortran: 18 first_arg
 [SUCCESS]   fortran: 19 fortran_std
 [SUCCESS]   fortran: 20 buildtype
  [ERROR]    fortran: 21 install static    (default_library=static)
During: build
Reason: Compiling source code failed.

Total configuration time: 29.29s
Total build time:         24.58s
Total test time:          0.52s

Total passed tests:  18
Total failed tests:  3
Total skipped tests: 0

Mesonlogs of failing tests


========================================================== fortran: 9 cpp ==========================================================
 

The Meson build system
Version: 0.59.99
Source dir: /fserver/mathewc/mat_meson/test cases/fortran/9 cpp
Build dir: /fserver/mathewc/mat_meson/b fd9bf17977
Build type: native build
Project name: C, C++ and Fortran
Project version: undefined
Fortran compiler for the host machine: nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
Fortran linker for the host machine: nagfor nag 7100
C compiler for the host machine: cc (gcc 11.1.1 "cc (GCC) 11.1.1 20210531 (Red Hat 11.1.1-3)")
C linker for the host machine: cc ld.bfd 2.35.1-41
C++ compiler for the host machine: c++ (gcc 11.1.1 "c++ (GCC) 11.1.1 20210531 (Red Hat 11.1.1-3)")
C++ linker for the host machine: c++ ld.bfd 2.35.1-41
Host machine cpu family: x86_64
Host machine cpu: x86_64
Build targets in project: 2

Found ninja-1.10.2 at /usr/bin/ninja
[1/8] Compiling C object cfort.p/main.c.o
[2/8] Module scanner.
[3/8] Module scanner.
[4/8] Compiling Fortran object cppfort.p/fortran.f.o
Obsolescent: ../test cases/fortran/9 cpp/fortran.f, line 1: Fixed source form
[5/8] Compiling Fortran object cfort.p/fortran.f.o
Obsolescent: ../test cases/fortran/9 cpp/fortran.f, line 1: Fixed source form
[6/8] Linking target cfort
FAILED: cfort 
cc  -o cfort cfort.p/main.c.o cfort.p/fortran.f.o -Wl,--as-needed -Wl,--no-undefined
/usr/bin/ld: cfort.p/fortran.f.o: in function `fortran':
/fserver/mathewc/mat_meson/b fd9bf17977/../test cases/fortran/9 cpp/fortran.f:7: undefined reference to `__NAGf90_random52_d'
/usr/bin/ld: /fserver/mathewc/mat_meson/b fd9bf17977/../test cases/fortran/9 cpp/fortran.f:9: undefined reference to `__NAGf90_pow'
collect2: error: ld returned 1 exit status
[7/8] Compiling C++ object cppfort.p/main.cpp.o
ninja: build stopped: subcommand failed.
 

ninja explain: deps for 'cfort.p/main.c.o' are missing
ninja explain: cfort.p/main.c.o is dirty
ninja explain: output cfort.p/depscan.dd doesn't exist
ninja explain: deps for 'cfort.p/fortran.f.o' are missing
ninja explain: cfort.p/fortran.f.o is dirty
ninja explain: cfort is dirty
ninja explain: output cppfort.p/depscan.dd doesn't exist
ninja explain: deps for 'cppfort.p/main.cpp.o' are missing
ninja explain: cppfort.p/main.cpp.o is dirty
ninja explain: deps for 'cppfort.p/fortran.f.o' are missing
ninja explain: cppfort.p/fortran.f.o is dirty
ninja explain: cppfort is dirty
ninja explain: loading dyndep file 'cppfort.p/depscan.dd'
ninja explain: output cppfort.p/main.cpp.o doesn't exist
ninja explain: output cppfort.p/fortran.f.o doesn't exist
ninja explain: cppfort.p/main.cpp.o is dirty
ninja explain: cppfort.p/fortran.f.o is dirty
ninja explain: cfort is dirty
ninja explain: cppfort is dirty
ninja explain: loading dyndep file 'cfort.p/depscan.dd'
ninja explain: output cfort.p/fortran.f.o doesn't exist
ninja explain: cfort.p/main.c.o is dirty
ninja explain: cfort.p/fortran.f.o is dirty
ninja explain: cfort is dirty
ninja explain: cppfort is dirty
 


======================================================== fortran: 15 include =======================================================
 

Build started at 2021-08-22T11:31:58.212087
Main binary: /fserver/mathewc/documents/Meson/nagfor/foo3/bin/python3
Build Options: -Dprefix=/usr -Dlibdir=lib -Dbackend=ninja
Python system: Linux
The Meson build system
Version: 0.59.99
Source dir: /fserver/mathewc/mat_meson/test cases/fortran/15 include
Build dir: /fserver/mathewc/mat_meson/b 342050c9e8
Build type: native build
Project name: Inclusive
Project version: undefined
Fortran compiler for the host machine: nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
Fortran linker for the host machine: nagfor nag 7100
Fortran compiler for the build machine: nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
Fortran linker for the build machine: nagfor nag 7100
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Adding test "Fortran include file hierarchy"
Adding test "Fortran include file syntax"
Running command: /usr/bin/ninja --version
--- stdout ---
1.10.2

--- stderr ---


Running command: /usr/bin/cmake --version
--- stdout ---
cmake version 3.20.5

CMake suite maintained and supported by Kitware (kitware.com/cmake).

--- stderr ---



Executing subproject cmake_inc method cmake 

CMake binary for 0 is not cached
CMake binary missing from cross or native file, or env var undefined.
Trying a default CMake fallback at cmake
Found CMake: /usr/bin/cmake (3.20.5)

Configuring the build directory with CMake version 3.20.5
Running CMake with: -G Ninja -DCMAKE_INSTALL_PREFIX=/usr
  - build directory:          /fserver/mathewc/mat_meson/b 342050c9e8/subprojects/cmake_inc/__CMake_build
  - source directory:         /fserver/mathewc/mat_meson/test cases/fortran/15 include/subprojects/cmake_inc
  - toolchain file:           /fserver/mathewc/mat_meson/b 342050c9e8/subprojects/cmake_inc/__CMake_build/CMakeMesonToolchainFile.cmake
  - preload file:             /fserver/mathewc/mat_meson/b 342050c9e8/meson-private/data/preload.cmake
  - trace args:               --trace-expand --trace-format=json-v1 --no-warn-unused-cli --trace-redirect=cmake_trace.txt
  - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089, CMP0102]

Calling CMake (['/usr/bin/cmake']) in /fserver/mathewc/mat_meson/b 342050c9e8/subprojects/cmake_inc/__CMake_build with:
  - "-G"
  - "Ninja"
  - "-DCMAKE_INSTALL_PREFIX=/usr"
  - "--trace-expand"
  - "--trace-format=json-v1"
  - "--no-warn-unused-cli"
  - "--trace-redirect=cmake_trace.txt"
  - "-DCMAKE_POLICY_WARNING_CMP0025=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0047=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0056=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0060=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0065=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0066=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0067=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0082=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0089=OFF"
  - "-DCMAKE_POLICY_WARNING_CMP0102=OFF"
  - "-DCMAKE_TOOLCHAIN_FILE=/fserver/mathewc/mat_meson/b 342050c9e8/subprojects/cmake_inc/__CMake_build/CMakeMesonToolchainFile.cmake"
  - "-DMESON_PRELOAD_FILE=/fserver/mathewc/mat_meson/b 342050c9e8/meson-private/data/preload.cmake"
  - "/fserver/mathewc/mat_meson/test cases/fortran/15 include/subprojects/cmake_inc"
Running with expanded trace output on.
Not searching for unused variables given on the command line.
Trace will be written to cmake_trace.txt
-- The Fortran compiler identification is NAG 7.1.7100
-- Detecting NAG Fortran directory
-- Detecting NAG Fortran directory - /fserver/nagprod/NP71/dev/NAG_Fortran-amd64-7100-295/lib
-- Detecting Fortran compiler ABI info
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_Fortran_PREPROCESS_SOURCE
CMake Error at /usr/share/cmake/Modules/CMakeDetermineCompilerABI.cmake:49 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  /usr/share/cmake/Modules/CMakeTestFortranCompiler.cmake:20 (CMAKE_DETERMINE_COMPILER_ABI)
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/fserver/mathewc/mat_meson/b 342050c9e8/subprojects/cmake_inc/__CMake_build/CMakeFiles/CMakeOutput.log".

CMake configuration: FAILED

test cases/fortran/15 include/meson.build:16:5: ERROR: Failed to configure the CMake subproject
 

Traceback (most recent call last):
  File "/fserver/mathewc/mat_meson/run_tests.py", line 283, in run_configure_inprocess
    returncode = mesonmain.run(commandlist, get_meson_script())
  File "/fserver/mathewc/mat_meson/mesonbuild/mesonmain.py", line 316, in run
    return CommandLineParser().run(args)
  File "/fserver/mathewc/mat_meson/mesonbuild/mesonmain.py", line 228, in run
    return options.run_func(options)
  File "/fserver/mathewc/mat_meson/mesonbuild/msetup.py", line 281, in run
    app.generate()
  File "/fserver/mathewc/mat_meson/mesonbuild/msetup.py", line 184, in generate
    self._generate(env)
  File "/fserver/mathewc/mat_meson/mesonbuild/msetup.py", line 223, in _generate
    intr.run()
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreter/interpreter.py", line 2374, in run
    super().run()
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 132, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 173, in evaluate_statement
    return self.evaluate_if(cur)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 257, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_statement
    return self.method_call(cur)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/interpreterbase.py", line 598, in method_call
    return self._holderify(obj.method_call(method_name, args, kwargs))
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreter/interpreterobjects.py", line 766, in method_call
    ret = method(state, args, kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/decorators.py", line 554, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/decorators.py", line 646, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/decorators.py", line 646, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/decorators.py", line 106, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreterbase/decorators.py", line 64, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/modules/cmake.py", line 394, in subproject
    subp = self.interpreter.do_subproject(dirname, 'cmake', kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreter/interpreter.py", line 812, in do_subproject
    raise e
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreter/interpreter.py", line 798, in do_subproject
    return self._do_subproject_cmake(subp_name, subdir, subdir_abs, default_options, kwargs)
  File "/fserver/mathewc/mat_meson/mesonbuild/interpreter/interpreter.py", line 875, in _do_subproject_cmake
    cm_int.initialise(cmake_options)
  File "/fserver/mathewc/mat_meson/mesonbuild/cmake/interpreter.py", line 939, in initialise
    cmake_exe = self.configure(extra_cmake_options)
  File "/fserver/mathewc/mat_meson/mesonbuild/cmake/interpreter.py", line 931, in configure
    raise CMakeException('Failed to configure the CMake subproject')
mesonbuild.cmake.common.CMakeException: Failed to configure the CMake subproject
 


====================================== fortran: 21 install static    (default_library=static) ======================================
 

The Meson build system
Version: 0.59.99
Source dir: /fserver/mathewc/mat_meson/test cases/fortran/21 install static
Build dir: /fserver/mathewc/mat_meson/b 0c301b5fbf
Build type: native build
Project name: try-static-subproject-dependency
Project version: undefined
Fortran compiler for the host machine: nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
Fortran linker for the host machine: nagfor nag 7100
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (1.7.3)
Found CMake: /usr/bin/cmake (3.20.5)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency static_hello found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency static_hello

Executing subproject static_hello 

static_hello| Project name: static-hello
static_hello| Project version: undefined
static_hello| Fortran compiler for the host machine: nagfor (nagfor 7100 "NAG Fortran Compiler Release 7.1(Hanzomon) Build 7100")
static_hello| Fortran linker for the host machine: nagfor nag 7100
static_hello| Build targets in project: 1
static_hello| Subproject static_hello finished.

Dependency static_hello from subproject subprojects/static_hello found: YES undefined
Build targets in project: 3

try-static-subproject-dependency undefined

  Subprojects
    static_hello: YES

Found ninja-1.10.2 at /usr/bin/ninja
[1/9] Module scanner.
[2/9] Module scanner.
[3/9] Module scanner.
[4/9] Compiling Fortran object subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o
[5/9] Linking static target subprojects/static_hello/libstatic_hello.a
[6/9] Compiling Fortran object libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o
[7/9] Linking static target libmainstatic.a
[8/9] Compiling Fortran object main_exe.p/main.f90.o
FAILED: main_exe.p/main.f90.o 
nagfor -Imain_exe.p -I. '-I../test cases/fortran/21 install static' -Ilibmainstatic.a.p -quiet -g -mdir main_exe.p -o main_exe.p/main.f90.o -c '../test cases/fortran/21 install static/main.f90'
Fatal Error: ../test cases/fortran/21 install static/main.f90, line 1: Cannot find module STATIC_HELLO
             detected at MAIN_LIB@<end-of-statement>
ninja: build stopped: subcommand failed.
 

ninja explain: output subprojects/static_hello/libstatic_hello.a.p/depscan.dd doesn't exist
ninja explain: deps for 'subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o' are missing
ninja explain: subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: output libmainstatic.a.p/depscan.dd doesn't exist
ninja explain: deps for 'libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o' are missing
ninja explain: libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: output main_exe.p/depscan.dd doesn't exist
ninja explain: deps for 'main_exe.p/main.f90.o' are missing
ninja explain: main_exe.p/main.f90.o is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: main_exe is dirty
ninja explain: loading dyndep file 'subprojects/static_hello/libstatic_hello.a.p/depscan.dd'
ninja explain: output subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o doesn't exist
ninja explain: subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o is dirty
ninja explain: output libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o doesn't exist
ninja explain: libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o is dirty
ninja explain: output main_exe.p/main.f90.o doesn't exist
ninja explain: main_exe.p/main.f90.o is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: main_exe is dirty
ninja explain: loading dyndep file 'main_exe.p/depscan.dd'
ninja explain: output main_exe.p/main.f90.o doesn't exist
ninja explain: main_exe.p/main.f90.o is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: main_exe is dirty
ninja explain: loading dyndep file 'libmainstatic.a.p/depscan.dd'
ninja explain: output libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o doesn't exist
ninja explain: libmainstatic.a.p/meson-generated_.._main_lib_output.F90.o is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a.p/meson-generated_.._static_hello_output.F90.o is dirty
ninja explain: output main_exe.p/main.f90.o doesn't exist
ninja explain: main_exe.p/main.f90.o is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: subprojects/static_hello/libstatic_hello.a is dirty
ninja explain: libmainstatic.a is dirty
ninja explain: main_exe is dirty
 

```

eli-schwartz commented on mesonbuild/compilers/fortran.py:
```diff
@@ -494,12 +494,32 @@ def __init__(self, exelist: T.List[str], version: str, for_machine: MachineChoic
         return ['-mdir', path]
 
+    @staticmethod
+    def get_nagfor_quiet(version) -> T.List[str]:
```
```suggestion
    def get_nagfor_quiet(version: str) -> T.List[str]:
```

eli-schwartz commented:
Great, CI passes now and even that mysterious " [UNEXRUN]   frameworks: 29 blocks" is gone now.

I'm afraid I have no clue about the other stuff, maybe someone else knows... or if it represents an overall improvement of support we can always just merge that as is, assuming someone who knows more about this section of the code is good with it.

eli-schwartz commented:
Rebased on master for general CI improvements and folded the lint fix into the commit that introduced the code.

matcross commented:
> I'm afraid I have no clue about the other stuff, maybe someone else knows... or if it represents an overall improvement of support we can always just merge that as is, assuming someone who knows more about this section of the code is good with it.

Ok, thanks. It's certainly not full support, you're right, but it does move from 'not working with nagfor at all' to having basic support seeming good. So I'm keen for this to go in as a first increment.

eli-schwartz commented:
Excellent, dcbaker gave his approval and I give my "this looks good to my non-expert eye" approval, so as soon as CI completes successfully again I'll merge this.

dcbaker commented:
Meson should be picking fortran for linking fortran and C code I think, but this is clearly a big improvement. If someone reminds me next week I can look at what happens with gfortran. I'm not sure about the other two issues either.

eli-schwartz commented:
> If someone reminds me next week

Ping. :D
